src/cascadia/TerminalControl/TSFInputControl.cpp

```diff
@@ -437,4 +437,26 @@ namespace winrt::Microsoft::Terminal::Control::implementation
     void TSFInputControl::_formatUpdatingHandler(CoreTextEditContext sender, const CoreTextFormatUpdatingEventArgs& /*args*/)
     {
     }
+
+    void TSFInputControl::ManuallyDisplayText(const winrt::hstring& text)
+    {
+        _focused = !text.empty();
+        Canvas().Visibility(text.empty() ? Visibility::Collapsed : Visibility::Visible);
+
+        _inputBuffer.clear();
+        _activeTextStart = 0;
+        _inComposition = false;
+
+        // HACK trim off leading DEL chars.
+        std::wstring_view view{ text.c_str() };
+        const auto strBegin = view.find_first_not_of(L"\x7f");
+        if (strBegin != std::wstring::npos)
+        {
+            view = view.substr(strBegin * 2);

```

Comment:
What we actually want to display is the text that would remain after accounting for the leading backspaces. So trim off the leading backspaces, AND and equal number of "real" characters.

(comment added)

Reply:
If we get an input like "remove 4 chars and it completes to foo", we turn that into `\x7f\x7f\x7f\x7ffoo`? And then TSF sees it and it writes...

Oh, while writing this comment, I just realized it would write a crashdump. 😄 There are only 7 chars in the `view` and it would try to `substr(8)`. There's no bounds check.

Why not just pass the remove amount to TSF? We already got the information for how much text needs to be removed.
Furthermore, does xterm.js consider the remove amount to be in extended grapheme clusters, code points or UTF16 code units? The `* 2` approach would only work with the last option. In other words, this should fail to work with surrogate pairs. You could use `til::utf16_iterate_next` as a stopgap.

Reply:
Huh, good point. I don't think that's really possible in practice (user types `abcd` -> autocomplete that to _something shorter??_), but it is a _protocol_ after all and there's no reason a dumb CLI app couldn't. 

> Why not just pass the remove amount to TSF

At this point, we've kinda abstracted a lot of info away in the interim. We've taken the replace count, turned it into backspaces and stitched that into the prefix of the text, and THEN asked the TermControl to preview {{backspaces + text}}. But TermControl (and sendInput) don't really know how to deal with "preview this text, and replace this many characters". Like, that's a weird mutation of `sendInput` (`sendInput(text="foo", replace=4)`)

> Furthermore, does xterm.js consider the remove amount

ho boy I'm sure that it's not specc'ed at all 😆 


----

Yea I might just back out the previewing.